### DISCUSSION TITLE
Why isn't importing a private module from a third-party library considered an error?

### DISCUSSION THREAD

**Discussion by Delgan** [microsoft/pyright#9004] [Q&A] [answered]
Hi.

I've read the docs about [Library interface](https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md#library-interface)  but I can't understand why isn't `pyright` reporting an error when I import private modules.

Consider first installing a third-party library like [`pydantic`](https://github.com/pydantic/pydantic/tree/v2.9.1/pydantic):

```bash
pip install pydantic==2.9.1
```

Then asking `pyright` to analyze the following file in "strict" mode:

```python
# pyright: strict
import pydantic._migration

print(pydantic._migration.MOVED_IN_V2)
```

Because I'm importing and using `_migration` which is supposed to be a private module, I would expect an error such as `reportPrivateUsage` or `reportPrivateImportUsage`. However, the output is (for version 1.1.380):

```bash
0 errors, 0 warnings, 0 informations 
```

Is that intentional design? I understand that for unit tests within the library, it's preferable for library authors to be able to import a private module (and test some internal functions). But shouldn't this be forbidden for a library user?

**Comment by erictraut** [comment, marked as answer]
Pyright doesn't check for this currently, but I agree it would be a good feature enhancement. Converting the discussion into an issue to track this.

**Reply by Hibou57** [reply]
Excuse me Sir, it is still a discussion, not an issue. But the original poster may repost as an issue, if he/she was invited to do.

**Comment by oliversheridanmethven** [comment]
There are some tools like flake8-self which try and do this, but none I can spot with the community momentum comparable to an established and popular type checker like this.

**Comment by Hibou57** [comment]
Unless I did not understand, a temporary work-around may be to use PyLint.
